core/src/main/java/org/apache/iceberg/BaseRewriteManifests.java

```diff
@@ -312,33 +311,62 @@ long getManifestTargetSizeBytes() {
   }
 
   class WriterWrapper {
-    private ManifestWriter writer;
-
-    synchronized void addEntry(ManifestEntry entry) {
-      if (writer == null) {
-        writer = newWriter();
-      } else if (writer.length() >= getManifestTargetSizeBytes()) {
-        close();
-        writer = newWriter();
+    private final Map<Integer, ManifestWriter> manifestWritersBySpecId = Collections.synchronizedMap(new HashMap<>());
+
+    synchronized void addEntry(ManifestEntry entry, int partitionSpecId) {
+      ManifestWriter manifestWriter = manifestWritersBySpecId.get(partitionSpecId);
+      if (manifestWriter == null) {
+        manifestWriter = newWriter(partitionSpecId);
+      } else if (manifestWriter.length() >= getManifestTargetSizeBytes()) {
+        close(partitionSpecId);
+        manifestWriter = newWriter(partitionSpecId);
       }
-      writer.existing(entry);
+      manifestWriter.existing(entry);
     }
 
-    private ManifestWriter newWriter() {
-      return new ManifestWriter(spec, manifestPath(manifestSuffix.getAndIncrement()), snapshotId());
+    private ManifestWriter newWriter(int partitionSpecId) {
+
+      if (manifestWritersBySpecId.containsKey(partitionSpecId)) {
+        // should not reach here, but checking to avoid resource leak.
+        throw new RuntimeException("ManifestWriter already exists for this partition-specId " + partitionSpecId);
```

Comment:
I usually like to start messages for "should not reach here" exceptions with "[BUG]: " so it's clear that it isn't the user's fault.

Also, "partition-specId" is odd because it mixes camelCase with hyphenated-names. How about changing it to "spec id"?